### DISCUSSION TITLE
Invalid value specified for 'add' operation

### DISCUSSION THREAD

**Discussion by Yunabell-VU** [eclipse-emfcloud/emfcloud#188] [ModelServer] [answered]
Hi, I'm working on a project which uses the EMFcloud as the server. Now I run into trouble with sending the ADD operation PATCH. 
Here's the context:  
In the server, I have an Ecore model called `statemachine.ecore` and its nsURI is `"http://nl.vu.cs.bumble/statemachine"`. 
The server also stores two different instances generated from the ecore model. One is called `StateMachine.xmi` and the other is `My.statemachine`. 
I tried to send Json Patches with the standard JSON pointer path, and it works on REPLACE and REMOVE for both instance models. However, I can not make a successful ADD patch to update `My.statemachine`. 

Here's the payload I send to the server:
```JSON
{
    "data": {
        "type": "modelserver.patch",
        "data": [{
                "op":"add",
                "path":"/input/0/",
                "value": {
		   "$type": "http://nl.vu.cs.bumble/statemachine#//Input",
                    "name":"test"
	        }
       }]
    }
}
```
And here is the response:
```JSON
{
    "type": "error",
    "data": "Invalid value specified for 'add' operation"
}
```
The weird thing is, the two instance models are the same, and this ADD patch applies successfully on the `StateMachine.xmi` model but it fails on the `My.statemachine` model.

Can you maybe give my some clues on why this is the case?

Thank you!

____________________________________________
An additional comment to the original question, `StateMachine.xmi` is manually added in the workspace, so it's already in there when the server is launched. However, `My.statemachine` is added to the server via a POST request. Is that the reason why the ADD operation does not work on `My.statemachine` model? How can I solve the problem?

**Comment by ndoschek** [comment, marked as answer]
Hey @Yunabell-VU,

thanks for your interest in the ModelServer and the EMF.cloud project! :tada: 

I assume the Json pointer path you are using (`"/input/0/"`) is either missing a feature name or you need to specify the path to the parent element if you want to add a new `Input` element to the model.

I can give you an example for the basic `coffeemodel` example we use in the `emfcloud-modelserver` project:

```
{
    "data": {
        "type": "modelserver.patch",
        "data": [
            {
                "op": "add",
                "path": "/workflows/0/nodes",
                "value": {
                    "$type": "http://www.eclipsesource.com/modelserver/example/coffeemodel#//AutomaticTask",
                    "name": "Brew"
                }
            }
        ]
    }
}
```

If we want to add a new node to the first `workflow` in the model, we define the Json pointer path like so: ` "path": "/workflows/0/nodes"`. This defines the first workflow in the model (`/workflow/0`) and appends the feature name `/nodes`, which is the container where the new element should be added.

On another note, you could also send the same patch using with EMF-like paths, which would lead to the same results:

```
...
                "path": "SuperBrewer3000.json#//@workflows.0/nodes",
...
```

HTH, and best regards from Vienna,
Nina

**Reply by Yunabell-VU** [reply]
Hi @ndoschek ,

Thank you very much for the reply.

I haven't figured out how to fix it yet, but we have another problem that also affects how we can write the payload of JSON patch. We are facing the issue that the models posted to the server are configured with all the `$type` represented in the local file system path instead of the model URI path.
E.g. The model URI for the model `My.statemachine` is `"http://nl.vu.cs.bumble/statemachine"` and it is added to the server through a POST request.

And this is how it looks on the server:
```JSON
 "data": {
        "$type": "file:/.../emfcloud-modelserver-master/examples/org.eclipse.emfcloud.modelserver.example/.temp/workspace/statemachine.ecore#//StateMachine",
        "$id": "/",
        "name": "statemachine",
        "input": [
            {
                "$id": "//@input.0",
                "name": "go",
                "baseconcept": [
                    {
                        "$type": "file:/.../emfcloud-modelserver-master/examples/org.eclipse.emfcloud.modelserver.example/.temp/workspace/statemachine.ecore#//Input",
                        "$id": "//@input.0/@baseconcept.0",
                        "name": ""
                    }
                ]
            }
}
```

Is there a way I can configure it so the `$type` of the model and its nodes can display the path with the model URI instead of the workspace file path? The file path is where the related ecore file is saved in the `resources/workspace` (The corresponding ecore model is added to the server manually by modifying the settings in `setupTempTestWorkspace()` provided by `ExampleServerLauncher`)

Best regards,
Yuna

**Reply by CamilleLetavernier** [reply]
Hi,

The $type represents the URI of the EMF EClass inside of your Ecore model. This URI will depend on how your model is created/generated.

If you use an EMF GenModel to generate Java classes from your Ecore Model, the URIs should be cleaner (In the form: <packageURI>#EClassName, where the Package URI is defined in your Ecore file). If you use a dynamic Ecore model (without generating Java classes), you have to change the way you load it during initialization (For example, using platform:/plugin/ scheme URI, if your Java model server is running in an Eclipse runtime. This scheme won't be available if you're running as a standard Java application, e.g. with a Main class. In that case, I'm not sure you can avoid the file:/ scheme with an absolute path)

When it comes to manipulating EMF Models, I would recommend the GenModel (Generating java classes) + Eclipse plug-in approach.
